File: SocketIoNodejs/canonical-client.js
const crypto = require('crypto');
const os = require('os');
const readline = require('readline');
const tls = require('tls');

const PROTOCOL_VERSION = '1.0';
const ALLOWED_COMMANDS = new Set(['health-check', 'refresh-config', 'collect-diagnostics', 'list-status', 'ping-time', 'custom-cmd']);

function createToken(secret, deviceId, role, lifetimeSeconds = 600) {
  const payload = {
    deviceId,
    role,
    expiresAtUtc: new Date(Date.now() + lifetimeSeconds * 1000).toISOString(),
    protocolVersion: PROTOCOL_VERSION,
  };
  const payloadBytes = Buffer.from(JSON.stringify(payload));
  const signature = crypto.createHmac('sha256', secret).update(payloadBytes).digest();
  return `${base64UrlEncode(payloadBytes)}.${base64UrlEncode(signature)}`;
}

function base64UrlEncode(buffer) {
  return buffer.toString('base64').replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/g, '');
}

class CanonicalTcpClient {
  constructor({ host, port, deviceId, role, authSecret, accessToken, allowUntrustedTls = false, timeoutMs = 2000, useLogin = false, loginUsername, loginPassword }) {
    this.host = host;
    this.port = port;
    this.deviceId = deviceId;
    this.role = role;
    this.authSecret = authSecret;
    this.accessToken = accessToken;
    this.allowUntrustedTls = allowUntrustedTls;
    this.timeoutMs = timeoutMs;
    this.useLogin = useLogin;
    this.loginUsername = loginUsername;
    this.loginPassword = loginPassword;
    this.socket = null;
    this.backlog = [];
    this.waiters = [];
    this.processedCommandIds = new Set();
  }

  async connect() {
    this.socket = await new Promise((resolve, reject) => {
      const socket = tls.connect({
        host: this.host,
        port: this.port,
        servername: 'localhost',
        rejectUnauthorized: !this.allowUntrustedTls,
      }, () => resolve(socket));
      socket.once('error', reject);
    });

    const lineReader = readline.createInterface({ input: this.socket });
    lineReader.on('line', (line) => {
      const envelope = JSON.parse(line);
      const waiterIndex = this.waiters.findIndex((waiter) => waiter.predicate(envelope));
      if (waiterIndex >= 0) {
        const [waiter] = this.waiters.splice(waiterIndex, 1);
        waiter.resolve(envelope);
        return;
      }

      this.backlog.push(envelope);
    });

    if (this.useLogin) {
      return this.login(this.loginUsername || this.deviceId, this.loginPassword || this.authSecret || '', this.role);
    }
    return this.authenticate();
  }

  close() {
    if (this.socket) {
      this.socket.destroy();
      this.socket = null;
    }
  }

  async authenticate() {
    const token = this.accessToken || createToken(this.authSecret || '', this.deviceId, this.role);
    const requestId = this.sendMessage('authenticate', { deviceId: this.deviceId, accessToken: token });
    const response = await this.waitFor((envelope) => envelope.requestId === requestId && ['authenticated', 'error'].includes(envelope.type));
    if (response.type === 'error') {
      throw new Error(response.payload.message);
    }
    return response;
  }

  async login(username, password, role = this.role) {
    const requestId = this.sendMessage('login', {
      deviceId: this.deviceId,
      username,
      password,
      role,
    });
    const response = await this.waitFor((envelope) => envelope.requestId === requestId && ['authenticated', 'error'].includes(envelope.type));
    if (response.type === 'error') {
      throw new Error(response.payload.message);
    }
    this.role = role;
    return response;
  }

  sendMessage(type, payload = null, { requestId, correlationId, deviceId, role } = {}) {
    if (!this.socket) {
      throw new Error('Client is not connected.');
    }

    const effectiveRequestId = requestId || crypto.randomUUID().replace(/-/g, '');
    const envelope = {
      protocolVersion: PROTOCOL_VERSION,
      requestId: effectiveRequestId,
      deviceId: deviceId || this.deviceId,
      clientId: deviceId || this.deviceId,
      role: role || this.role,
      type,
      correlationId: correlationId || null,
      timestampUtc: new Date().toISOString(),
      payload,
    };
    this.socket.write(`${JSON.stringify(envelope)}\n`);
    return effectiveRequestId;
  }

  async readMessage(timeoutMs = this.timeoutMs) {
    if (this.backlog.length > 0) {
      return this.backlog.shift();
    }

    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        const index = this.waiters.indexOf(waiter);
        if (index >= 0) {
          this.waiters.splice(index, 1);
        }
        reject(new Error('Timed out waiting for a matching message.'));
      }, timeoutMs);

      const waiter = {
        predicate: () => true,
        resolve: (envelope) => {
          clearTimeout(timeout);
          resolve(envelope);
        },
      };
      this.waiters.push(waiter);
    });
  }

  async waitFor(predicate, timeoutMs = this.timeoutMs) {
    const backlogIndex = this.backlog.findIndex(predicate);
    if (backlogIndex >= 0) {
      return this.backlog.splice(backlogIndex, 1)[0];
    }

    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        const index = this.waiters.indexOf(waiter);
        if (index >= 0) {
          this.waiters.splice(index, 1);
        }
        reject(new Error('Timed out waiting for a matching message.'));
      }, timeoutMs);

      const waiter = {
        predicate,
        resolve: (envelope) => {
          clearTimeout(timeout);
          resolve(envelope);
        },
      };
      this.waiters.push(waiter);
    });
  }

  async sendAdminCommand(commandName, { commandId, targetMode = 'all', targetDeviceIds = [], timeoutMs = 1000 } = {}) {
    const requestId = this.sendMessage('admin-command', {
      commandId: commandId || crypto.randomUUID().replace(/-/g, ''),
      commandName,
      targetMode,
      targetDeviceIds,
      timeoutMs,
    });
    return this.waitFor((envelope) => envelope.requestId === requestId && ['admin-command.accepted', 'error'].includes(envelope.type), timeoutMs + 1000);
  }

  async processNextCommand() {
    const command = await this.waitFor((envelope) => envelope.type === 'admin-command', 5000);
    const { commandId, commandName } = command.payload;
    this.sendMessage('command-ack', { commandId, status: 'accepted' }, { correlationId: commandId });

    const duplicate = this.processedCommandIds.has(commandId);
    if (!duplicate) {
      this.processedCommandIds.add(commandId);
    }

    const resultPayload = {
      commandId,
      commandName,
      status: 'completed',
      success: true,
      duplicate,
      result: this.executeCommand(commandName),
    };
    this.sendMessage('command-result', resultPayload, { correlationId: commandId });
    return { command, resultPayload };
  }

  executeCommand(commandName) {
    if (!ALLOWED_COMMANDS.has(commandName)) {
      throw new Error(`Command '${commandName}' is not allowlisted.`);
    }

    if (commandName === 'health-check') {
      return { status: 'ok', deviceId: this.deviceId, observedAtUtc: new Date().toISOString() };
    }

    if (commandName === 'refresh-config') {
      return { status: 'refreshed', deviceId: this.deviceId, appliedAtUtc: new Date().toISOString() };
    }

    if (commandName === 'collect-diagnostics') {
      return { status: 'collected', deviceId: this.deviceId, diagnostics: { platform: os.platform(), pid: process.pid } };
    }

    if (commandName === 'list-status') {
      return {
        status: 'ready',
        deviceId: this.deviceId,
        role: this.role,
        platform: os.platform(),
        observedAtUtc: new Date().toISOString(),
      };
    }

    if (commandName === 'ping-time') {
      return { status: 'pong', deviceId: this.deviceId, observedAtUtc: new Date().toISOString() };
    }

    if (commandName === 'custom-cmd') {
      return {
        status: 'acknowledged',
        executed: false,
        deviceId: this.deviceId,
        note: 'allowlisted stub only; no shell or code execution',
        observedAtUtc: new Date().toISOString(),
      };
    }

    throw new Error(`Command '${commandName}' is not implemented.`);
  }
}

function parseArgs(argv) {
  const result = { host: '127.0.0.1', targetDeviceIds: [] };
  for (let index = 2; index < argv.length; index += 1) {
    const key = argv[index];
    const value = argv[index + 1];
    switch (key) {
      case '--host':
        result.host = value;
        index += 1;
        break;
      case '--port':
        result.port = Number(value);
        index += 1;
        break;
      case '--device-id':
        result.deviceId = value;
        index += 1;
        break;
      case '--role':
        result.role = value;
        index += 1;
        break;
      case '--auth-secret':
        result.authSecret = value;
        index += 1;
        break;
      case '--access-token':
        result.accessToken = value;
        index += 1;
        break;
      case '--send-command':
        result.sendCommand = value;
        index += 1;
        break;
      case '--target-mode':
        result.targetMode = value;
        index += 1;
        break;
      case '--target-device-id':
        result.targetDeviceIds.push(value);
        index += 1;
        break;
      case '--timeout-ms':
        result.timeoutMs = Number(value);
        index += 1;
        break;
      case '--allow-untrusted':
        result.allowUntrustedTls = true;
        break;
      case '--login':
        result.useLogin = true;
        break;
      case '--login-user':
        result.loginUsername = value;
        index += 1;
        break;
      case '--login-password':
        result.loginPassword = value;
        index += 1;
        break;
      default:
        break;
    }
  }
  return result;
}

async function main() {
  const args = parseArgs(process.argv);
  const client = new CanonicalTcpClient({
    host: args.host,
    port: args.port,
    deviceId: args.deviceId,
    role: args.role,
    authSecret: args.authSecret || process.env.SOCKET_PLAYGROUND_AUTH_SECRET,
    accessToken: args.accessToken,
    allowUntrustedTls: !!args.allowUntrustedTls,
    timeoutMs: args.timeoutMs || 2000,
    useLogin: !!args.useLogin,
    loginUsername: args.loginUsername,
    loginPassword: args.loginPassword,
  });

  try {
    const authenticated = await client.connect();
    console.log(JSON.stringify(authenticated));

    if (args.sendCommand) {
      const accepted = await client.sendAdminCommand(args.sendCommand, {
        targetMode: args.targetMode || 'all',
        targetDeviceIds: args.targetDeviceIds,
        timeoutMs: args.timeoutMs || 1000,
      });
      console.log(JSON.stringify(accepted));
      const commandId = accepted.payload.commandId;
      while (true) {
        const envelope = await client.readMessage(args.timeoutMs || 2000);
        console.log(JSON.stringify(envelope));
        if (envelope.type === 'command-summary' && envelope.correlationId === commandId) {
          break;
        }
      }
    } else {
      console.log('Connected; waiting for allowlisted admin-command messages. Press Ctrl+C to exit.');
      while (true) {
        const { command, resultPayload } = await client.processNextCommand();
        console.log(JSON.stringify(command));
        console.log(JSON.stringify(resultPayload));
      }
    }
  } finally {
    client.close();
  }
}

if (require.main === module) {
  main().catch((error) => {
    console.error(error);
    process.exitCode = 1;
  });
}

module.exports = {
  CanonicalTcpClient,
  createToken,
};

File: SocketIoServerPython/canonical_client.py
import argparse
import base64
import hashlib
import hmac
import json
import os
import platform
import socket
import ssl
import time
import uuid
from datetime import datetime, timedelta, timezone


PROTOCOL_VERSION = "1.0"
ALLOWED_COMMANDS = {"health-check", "refresh-config", "collect-diagnostics", "list-status", "ping-time"}


def _utc_now():
    return datetime.now(timezone.utc)


def _base64url_encode(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).rstrip(b"=").decode("ascii")


def create_token(secret: str, device_id: str, role: str, lifetime_seconds: int = 600) -> str:
    payload = {
        "deviceId": device_id,
        "role": role,
        "expiresAtUtc": (_utc_now() + timedelta(seconds=lifetime_seconds)).isoformat(),
        "protocolVersion": PROTOCOL_VERSION,
    }
    payload_bytes = json.dumps(payload, separators=(",", ":"), sort_keys=True).encode("utf-8")
    signature = hmac.new(secret.encode("utf-8"), payload_bytes, hashlib.sha256).digest()
    return f"{_base64url_encode(payload_bytes)}.{_base64url_encode(signature)}"


class CanonicalTcpClient:
    def __init__(
        self,
        host: str,
        port: int,
        device_id: str,
        role: str,
        auth_secret: str | None = None,
        access_token: str | None = None,
        allow_untrusted_tls: bool = False,
        timeout_seconds: float = 2.0,
    ) -> None:
        self.host = host
        self.port = port
        self.device_id = device_id
        self.role = role
        self.auth_secret = auth_secret
        self.access_token = access_token
        self.allow_untrusted_tls = allow_untrusted_tls
        self.timeout_seconds = timeout_seconds
        self._socket: ssl.SSLSocket | None = None
        self._reader = None
        self._writer = None
        self._backlog: list[dict] = []
        self._processed_command_ids: set[str] = set()

    def connect(self) -> dict:
        raw_socket = socket.create_connection((self.host, self.port), timeout=self.timeout_seconds)
        raw_socket.settimeout(self.timeout_seconds)
        if self.allow_untrusted_tls:
            context = ssl._create_unverified_context()
        else:
            context = ssl.create_default_context()

        self._socket = context.wrap_socket(raw_socket, server_hostname="localhost")
        self._reader = self._socket.makefile("r", encoding="utf-8", newline="\n")
        self._writer = self._socket.makefile("w", encoding="utf-8", newline="\n")
        return self.authenticate()

    def close(self) -> None:
        if self._writer is not None:
            self._writer.close()
        if self._reader is not None:
            self._reader.close()
        if self._socket is not None:
            self._socket.close()
        self._writer = None
        self._reader = None
        self._socket = None

    def authenticate(self) -> dict:
        token = self.access_token or create_token(self.auth_secret or "", self.device_id, self.role)
        request_id = self.send_message("authenticate", {"deviceId": self.device_id, "accessToken": token})
        response = self.wait_for(lambda envelope: envelope["requestId"] == request_id and envelope["type"] in {"authenticated", "error"})
        if response["type"] == "error":
            raise RuntimeError(response["payload"]["message"])
        return response

    def send_message(
        self,
        message_type: str,
        payload: dict | None = None,
        *,
        request_id: str | None = None,
        correlation_id: str | None = None,
        device_id: str | None = None,
        role: str | None = None,
    ) -> str:
        if self._writer is None:
            raise RuntimeError("Client is not connected.")

        request_id = request_id or uuid.uuid4().hex
        envelope = {
            "protocolVersion": PROTOCOL_VERSION,
            "requestId": request_id,
            "deviceId": device_id or self.device_id,
            "clientId": device_id or self.device_id,
            "role": role or self.role,
            "type": message_type,
            "correlationId": correlation_id,
            "timestampUtc": _utc_now().isoformat(),
            "payload": payload,
        }
        self._writer.write(json.dumps(envelope, separators=(",", ":")) + "\n")
        self._writer.flush()
        return request_id

    def wait_for(self, predicate, timeout_seconds: float | None = None) -> dict:
        deadline = time.monotonic() + (timeout_seconds or self.timeout_seconds)
        backlog: list[dict] = []
        while time.monotonic() < deadline:
            envelope = self.read_message(max(0.1, deadline - time.monotonic()))
            if predicate(envelope):
                for item in reversed(backlog):
                    self._push_back(item)
                return envelope
            backlog.append(envelope)
        for item in reversed(backlog):
            self._push_back(item)
        raise TimeoutError("Timed out waiting for a matching message.")

    def _push_back(self, envelope: dict) -> None:
        self._backlog.insert(0, envelope)

    def _pop_backlog(self):
        if self._backlog:
            return self._backlog.pop(0)
        return None

    def read_message(self, timeout_seconds: float | None = None) -> dict:
        queued = self._pop_backlog()
        if queued is not None:
            return queued

        if self._socket is None or self._reader is None:
            raise RuntimeError("Client is not connected.")

        previous_timeout = self._socket.gettimeout()
        self._socket.settimeout(timeout_seconds or self.timeout_seconds)
        try:
            line = self._reader.readline()
        finally:
            self._socket.settimeout(previous_timeout)

        if not line:
            raise RuntimeError("Connection closed by server.")
        return json.loads(line)

    def send_admin_command(self, command_name: str, *, command_id: str | None = None, target_mode: str = "all", target_device_ids=None, timeout_ms: int = 1000) -> dict:
        request_id = self.send_message(
            "admin-command",
            {
                "commandId": command_id or uuid.uuid4().hex,
                "commandName": command_name,
                "targetMode": target_mode,
                "targetDeviceIds": target_device_ids or [],
                "timeoutMs": timeout_ms,
            },
        )
        return self.wait_for(lambda envelope: envelope["requestId"] == request_id and envelope["type"] in {"admin-command.accepted", "error"})

    def process_next_command(self) -> tuple[dict, dict]:
        command = self.wait_for(lambda envelope: envelope["type"] == "admin-command", timeout_seconds=5.0)
        payload = command["payload"]
        command_id = payload["commandId"]
        command_name = payload["commandName"]
        self.send_message("command-ack", {"commandId": command_id, "status": "accepted"}, correlation_id=command_id)

        duplicate = command_id in self._processed_command_ids
        if not duplicate:
            self._processed_command_ids.add(command_id)

        result_payload = {
            "commandId": command_id,
            "commandName": command_name,
            "status": "completed",
            "success": True,
            "duplicate": duplicate,
            "result": self._execute_command(command_name),
        }
        self.send_message("command-result", result_payload, correlation_id=command_id)
        return command, result_payload

    def _execute_command(self, command_name: str) -> dict:
        if command_name not in ALLOWED_COMMANDS:
            raise RuntimeError(f"Command '{command_name}' is not allowlisted.")

        if command_name == "health-check":
            return {"status": "ok", "deviceId": self.device_id, "observedAtUtc": _utc_now().isoformat()}
        if command_name == "refresh-config":
            return {"status": "refreshed", "deviceId": self.device_id, "appliedAtUtc": _utc_now().isoformat()}
        if command_name == "collect-diagnostics":
            return {
                "status": "collected",
                "deviceId": self.device_id,
                "diagnostics": {"platform": platform.platform(), "pid": os.getpid()},
            }
        if command_name == "list-status":
            return {
                "status": "ready",
                "deviceId": self.device_id,
                "role": self.role,
                "platform": platform.platform(),
                "observedAtUtc": _utc_now().isoformat(),
            }
        if command_name == "ping-time":
            return {"status": "pong", "deviceId": self.device_id, "observedAtUtc": _utc_now().isoformat()}
        raise RuntimeError(f"Command '{command_name}' is not implemented.")


def _parse_args():
    parser = argparse.ArgumentParser(description="Canonical raw TCP/TLS client for SocketPlayground.")
    parser.add_argument("--host", default="127.0.0.1")
    parser.add_argument("--port", type=int, required=True)
    parser.add_argument("--device-id", required=True)
    parser.add_argument("--role", choices=["admin", "client"], required=True)
    parser.add_argument("--auth-secret", default=os.getenv("SOCKET_PLAYGROUND_AUTH_SECRET"))
    parser.add_argument("--access-token")
    parser.add_argument("--allow-untrusted", action="store_true")
    parser.add_argument("--send-command")
    parser.add_argument("--target-mode", choices=["all", "devices"], default="all")
    parser.add_argument("--target-device-id", action="append", default=[])
    parser.add_argument("--timeout-ms", type=int, default=1000)
    return parser.parse_args()


def main():
    args = _parse_args()
    client = CanonicalTcpClient(
        args.host,
        args.port,
        args.device_id,
        args.role,
        auth_secret=args.auth_secret,
        access_token=args.access_token,
        allow_untrusted_tls=args.allow_untrusted,
        timeout_seconds=max(2.0, args.timeout_ms / 1000),
    )

    try:
        authenticated = client.connect()
        print(json.dumps(authenticated))

        if args.send_command:
            accepted = client.send_admin_command(
                args.send_command,
                target_mode=args.target_mode,
                target_device_ids=args.target_device_id,
                timeout_ms=args.timeout_ms,
            )
            print(json.dumps(accepted))
            command_id = accepted["payload"]["commandId"]
            while True:
                envelope = client.read_message(timeout_seconds=max(2.0, args.timeout_ms / 1000))
                print(json.dumps(envelope))
                if envelope["type"] == "command-summary" and envelope.get("correlationId") == command_id:
                    break
        else:
            print("Connected; waiting for allowlisted admin-command messages. Press Ctrl+C to exit.")
            while True:
                command, result = client.process_next_command()
                print(json.dumps(command))
                print(json.dumps(result))
    except KeyboardInterrupt:
        pass
    finally:
        client.close()


if __name__ == "__main__":
    main()
